### PR TITLE
Adding GCS artifact storage capabilities.

### PR DESCRIPTION
**Pull request by bnekolny**

### CODE REVIEW

**Comment by bnekolny** [comment]
I still need to add some tests (working through the nuances of no `moto` equivalent for GCS.

We're using GCS for storage, so is there an appetite for this being merged into master?

**Comment by mateiz** [comment]
This definitely sounds like a good idea as long as you provide a way to test it. It's not a lot of code, and it will obviously help a lot of users.

One question about the URI scheme: is the gs:// scheme also recognized for Google Cloud Storage in TensorFlow, Spark, and other systems that support writing to cloud storage? Or are multiple schemes used? In the latter case, we'll probably want to choose the most common one because we want to make it possible for jobs to write directly to the artifact URI for a run without writing to a local file first (especially for large datasets). We already have this issue with S3, where Hadoop/Spark use s3a or s3n, and we plan to do the conversion automatically there.

**Comment by bnekolny** [comment]
@mateiz the gs:// scheme is a standard reference for GCS, much like s3://. There are different HTTP urls, such as the S3 equivalent of s3.amazonaws.com, but I believe gs:// is more appropriate for a storage mechanism.

I added some tests, but let me know your thoughts. We're running on GCP and are already using a fork with this code in it, but are hoping to get it merged into mainline.

**Comment by smurching** [comment]
Ah this is awesome, thanks @bnekolny - giving it a look now :)

**Review by smurching — COMMENTED** [review]
Thanks a ton for the PR @bnekolny, this will definitely be helpful for users - leaving a few comments :)

File: mlflow/store/artifact_repo.py [hunk 1]
```diff
@@ -186,3 +189,89 @@ def _download_artifacts_into(self, artifact_path, dest_dir):
...
+
+    def list_artifacts(self, path=None):
+        (bucket, artifact_path) = self.parse_gcs_uri(self.artifact_uri)
+        dest_path = artifact_path
+        if path:
+            dest_path = build_path(dest_path, path)
+        infos = []
+        prefix = dest_path + "/"
+
+        results = self.gcs.Client().get_bucket(bucket).list_blobs(prefix=prefix)
```
[the remark this answers is not in the document]
Comment (bnekolny) [reply mlflow/store/artifact_repo.py]: I can confirm that passing no page_token does return all the objects. I haven't looked at the implementation, but I've gotten >50k objects back from the python sdk without using page_tokens.

File: mlflow/store/artifact_repo.py [hunk 2]
```diff
@@ -186,3 +189,89 @@ def _download_artifacts_into(self, artifact_path, dest_dir):
             s3_path = build_path(s3_path, artifact_path)
             boto3.client('s3').download_file(bucket, s3_path, local_path)
         return local_path
+
+
+class GCSArtifactRepository(ArtifactRepository):
+    """Stores artifacts on Google Cloud Storage.
```
[the remark this answers is not in the document]
Comment (bnekolny) [reply mlflow/store/artifact_repo.py]: I added docs here: https://github.com/databricks/mlflow/pull/152/commits/e387bd0e6034683330305ad447104fd5258f3fad

**Comment by bnekolny** [comment]
@smurching I've updated the code to address all of your comments, let me know if there is anything else. I'll watch to make sure tests pass and tweak things if necessary to make sure those go through.

**Review by smurching — APPROVED** [review]
Awesome this LGTM, merging to master - thanks for the hard work on this @bnekolny :)

We're hoping to make a release in the next few days so you should see the GCS functionality in the PyPi installation of MLflow soon!

**Pull request merged**